Feature/us16/profil

Martin-Hayot opened the pull request:
# Description de la US
Cet US consiste à créer une page de profil qui permet de changer une photo de profil, et ses informations personnelles tout en redemandant une authentification pour sécuriser l'utilisateur.

lien us : https://github.com/users/Martin-Hayot/projects/2?pane=issue&itemId=42738463

## Modification
### Frontend
- Nécessite l'authentification pour pouvoir y accéder
- Une page contenant la possibilité de changer sa photo de profil et un formulaire pour les informations personnelles.
- Un popup s'affiche lorsqu'on clique sur apply changes (appliquer les modifs) qui nous demandent nos informations de connexion.
- Lorsque l'utilisateur est authentifié, affiche un Toast en fonction de si tout c'est bien passé ou s'il y a eu une erreur.

### Backend
- Ajout d'une route user contenant 3 nouveau endpoint
- put "/profile/picture"
- get "/profile/picture"
- patch "/profile"
- Nouvelle entrée dans la db : profilePicture

Julien00859 commented:
You're asking me a lot of reviews and frankly I just cannot keep pace, can you instead select which PR you actually want my review and ping me at most once a week? I'm also interested to see how you review each other. I would suggest you to ping me at the very end, when you're confident the PR is ready to be merge. Unsubscribing for now, ping me at the end :p

JarodEphec reviewed (changes requested):
Salut Martin, très bon travail. Si j'avais su que l'on pouvait enregistrer des images comme ça en DB, ça m'aurait sauvé de quelques soucis en DEV3, enfin, j'ai juste quelques remarques à faire.

- Le changement de password ne fonctionne pas, ce qui est bizarre, car la vérification entre le nouveau et l'ancien password lui fonctionne.  
- C'est un peu bizarre que quand on va sur la page, on me demande mon email, je trouve que cette information devrait être rentrée par défaut et que dans le modal, il ne soit pas possible de la changer, parce que là, on demande les anciennes infos, je ne sais pas si tu me comprends ? Une fois dans le modal, ça serait peut-être mieux de mettre une indication comme quoi, on demande les informations de connexion avant changement.
- Je peux me rendre sur la page sans être connecté, j'imagine que pour ce problème, tu attends d'avoir les `UserContextes` de Samuel ?
- Un petit test de render en front ne ferait pas de mal.

JarodEphec commented on g-sonde/src/pages/User/Profile.tsx:
```diff
@@ -0,0 +1,326 @@
+    };
+
+    const handleMainFormSubmit = () => {
+        axios
+            .patch(
+                "http://localhost:3001/api/user/profile",
+                {
+                    email: email,
+                    firstname: firstname,
+                    lastname: lastname,
+                    password: password,
+                },
+                {
+                    withCredentials: true,
+                }
+            )
```
Nouveau mdp envoyé en clair ?

Martin-Hayot replied:
Oui c'est comme pour la création de compte, on est obligé de faire comme ça

JarodEphec commented on g-sonde/src/pages/User/Profile.tsx:
```diff
@@ -0,0 +1,326 @@
+                                        <IonLabel position="floating">
+                                            Firstname
+                                        </IonLabel>
+                                        <IonInput
+                                            type="email"
+                                            onIonChange={(e) =>
+                                                setFirstname(e.detail.value!)
+                                            }
+                                            required
+                                        ></IonInput>
```
[Ionic Warning]: ion-input now requires providing a label with either the "label" property or the "aria-label" attribute. To migrate, remove any usage of "ion-label" and pass the label text to either the "label" property or the "aria-label" attribute.

Example: <ion-input label="Email"></ion-input>
Example with aria-label: <ion-input aria-label="Email"></ion-input>

For inputs that do not render the label immediately next to the input, developers may continue to use "ion-label" but must manually associate the label with the input by using "aria-labelledby".

Developers can use the "legacy" property to continue using the legacy form markup. This property will be removed in an upcoming major release of Ionic where this form control will use the modern form markup.

JarodEphec commented on backend/routes/user.js:
```diff
@@ -0,0 +1,86 @@
+            firstname,
+            lastname,
+            email,
+            hashedPassword,
```
Je pense avoir trouvé le fix pour le changement de password : 
```
password: hashedPassword,
```

Martin-Hayot replied:
je comprends pas ce que tu veux dire par ça. Pourrais tu me donner plus de contexte ?

JarodEphec replied:
Ce qui fait que le password ne se change pas, avec ce que j'ai fait comme changement ici, il se change bien chez moi.

Martin-Hayot replied:
oui mais je ne comprends pas ou tu rajoute la ligne en question.

```js
const hashedPassword = await bcrypt.hash(password, 10);
const updateUser = await User.findByIdAndUpdate(userId, {
            firstname,
            lastname,
            email,
            password: hashedPassword,
        });
```

c'est ça qui marche chez toi ?

Martin-Hayot commented:
> Salut Martin, très bon travail. Si j'avais su que l'on pouvait enregistrer des images comme ça en DB, ça m'aurait sauvé de quelques soucis en DEV3, enfin, j'ai juste quelques remarques à faire.
> 
> * Le changement de password ne fonctionne pas, ce qui est bizarre, car la vérification entre le nouveau et l'ancien password lui fonctionne.
> * C'est un peu bizarre que quand on va sur la page, on me demande mon email, je trouve que cette information devrait être rentrée par défaut et que dans le modal, il ne soit pas possible de la changer, parce que là, on demande les anciennes infos, je ne sais pas si tu me comprends ? Une fois dans le modal, ça serait peut-être mieux de mettre une indication comme quoi, on demande les informations de connexion avant changement.
> * Je peux me rendre sur la page sans être connecté, j'imagine que pour ce problème, tu attends d'avoir les `UserContextes` de Samuel ?
> * Un petit test de render en front ne ferait pas de mal.

Pour le password je vais voir pourquoi il ne se modifie pas. Le formulaire est vide de base parce que toute les informations doivent être modifiable même l'email si jamais il veut changer l'endroit où arrive les notifications. Je vais rajouter un test e2e et modifier un peu la page pour un meilleur UX

JarodEphec commented:
> > Salut Martin, très bon travail. Si j'avais su que l'on pouvait enregistrer des images comme ça en DB, ça m'aurait sauvé de quelques soucis en DEV3, enfin, j'ai juste quelques remarques à faire.
> > 
> > * Le changement de password ne fonctionne pas, ce qui est bizarre, car la vérification entre le nouveau et l'ancien password lui fonctionne.
> > * C'est un peu bizarre que quand on va sur la page, on me demande mon email, je trouve que cette information devrait être rentrée par défaut et que dans le modal, il ne soit pas possible de la changer, parce que là, on demande les anciennes infos, je ne sais pas si tu me comprends ? Une fois dans le modal, ça serait peut-être mieux de mettre une indication comme quoi, on demande les informations de connexion avant changement.
> > * Je peux me rendre sur la page sans être connecté, j'imagine que pour ce problème, tu attends d'avoir les `UserContextes` de Samuel ?
> > * Un petit test de render en front ne ferait pas de mal.
> 
> Pour le password je vais voir pourquoi il ne se modifie pas. Le formulaire est vide de base parce que toute les informations doivent être modifiable même l'email si jamais il veut changer l'endroit où arrive les notifications. Je vais rajouter un test e2e et modifier un peu la page pour un meilleur UX

Oui mais je trouve que tout ce qui est information comme le nom, prénom et adresse email devrait être prérempli.

Martin-Hayot committed:
US16 feat: page de profile
Description de la US
Cet US consiste à créer une page de profil qui permet de changer une photo de profil, et ses informations personnelles tout en redemandant une authentification pour sécuriser l'utilisateur.

lien us : https://github.com/users/Martin-Hayot/projects/2?pane=issue&itemId=42738463

Modification
Frontend
Nécessite l'authentification pour pouvoir y accéder
Une page contenant la possibilité de changer sa photo de profil et un formulaire pour les informations personnelles.
Un popup s'affiche lorsqu'on clique sur apply changes (appliquer les modifs) qui nous demandent nos informations de connexion.
Lorsque l'utilisateur est authentifié, affiche un Toast en fonction de si tout c'est bien passé ou s'il y a eu une erreur.
Backend
Ajout d'une route user contenant 3 nouveau endpoint
put "/profile/picture"
get "/profile/picture"
patch "/profile"
Nouvelle entrée dans la db : profilePicture

JarodEphec reviewed (approved):
Parfait, toutes mes remarques ont été prises en compte, good job !
